### PR TITLE
Adding Geant4, migration from HomeBrew/Science

### PR DESCRIPTION
**Pull request by henryiii**
The old Geant4 formula in the science tap was beginning to have problems. I've updated the formula for the latest version of Geant4 from CERN, and fixed the file download issues. If you add `--with-qt`, you can get the optional GUI, as well.

### CODE REVIEW

File: Formula/geant4.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
+class Geant4 < Formula
+  desc "Simulation toolkit for particle transport through matter"
+  homepage "http://geant4.cern.ch"
+  url "http://geant4.cern.ch/support/source/geant4.10.04.p02.tar.gz"
+  version "10.4.2"
+  sha256 "2cac09e799f2eb609a7f1e4082d3d9d3d4d9e1930a8c4f9ecdad72aad35cdf10"
+
+  option "with-g3tog4", "Use G3toG4 Library"
```
Comment (fxcoudert) [inline Formula/geant4.rb]: Please remove all options, choosing a "best use case" scenario for most users. No experimental code (unless upstream enables it by default).

File: Formula/geant4.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  option "with-usolids", "Use USolids (experimental)"
+  option "without-multithreaded", "Build without multithreading support"
+
+  depends_on "cmake"
+  depends_on :x11
+  depends_on "xerces-c" => :recommended
```
Comment (fxcoudert) [inline Formula/geant4.rb]: Make it mandatory.

File: Formula/geant4.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      args << "-DGEANT4_BUILD_MULTITHREADED=ON" if build.with? "multithreaded"
+      args.concat(std_cmake_args)
+      system "cmake", *args
+      system "make", "install"
+    end
+  end
+
+  def post_install
+    resources.each do |r|
+      (share/"Geant4-#{version}/data/#{r.name}#{r.version}").install r
```
Comment (fxcoudert) [inline Formula/geant4.rb]: Why is this in postinstall and not in install?

File: Formula/geant4.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  end
+
+  test do
+    system "cmake", share/"Geant4-#{version}/examples/basic/B1"
+    system "make"
+    (testpath/"test.sh").write <<~EOS
+      . #{bin}/geant4.sh
+      ./exampleB1 run2.mac
+    EOS
+    system "/bin/bash", "test.sh"
```
Comment (fxcoudert) [inline Formula/geant4.rb]: Can we check the output?

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/geant4.rb (hunk 3 above)
Comment (henryiii) [reply Formula/geant4.rb]: I believe in post install they don't become part of the bottle (one or two of them are large), and you can also rerun the post-install without rebuilding (that's how I was working on developing it). If there's a reason to move it, I can.

File: Formula/geant4.rb (hunk 2 above)
Comment (henryiii) [reply Formula/geant4.rb]: Fine by me. :)

File: Formula/geant4.rb (hunk 1 above)
Comment (henryiii) [reply Formula/geant4.rb]: Okay. I don't think anyone uses G3toG4 anymore, though I'm not aware of a downside to including it, so I'll remove it. I'll remove the other options. I'll leave the Qt GUI in as optional.

**Comment by fxcoudert** [comment]
Given the current analytics:

```
install events in the last 30 days
================================================================================
1 | brewsci/science/geant4                                        |  7 |  50.00%
2 | brewsci/science/geant4 --with-qt                              |  4 |  28.57%
3 | brewsci/science/geant4 --with-gdml                            |  2 |  14.29%
4 | brewsci/science/geant4 --with-gdml --with-qt                  |  1 |   7.14%
================================================================================
Total                                                             | 14 |    100%
================================================================================
```

it makes sense to include Qt by default. No options, especially for formulas with relatively few users.

**Comment by henryiii** [comment]
I've added Qt as a required dependency, and I removed the X11 dependency (Qt doesn't need X11). I also added a Caveat to help users.

File: Formula/geant4.rb [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  url "http://geant4.cern.ch/support/source/geant4.10.04.p02.tar.gz"
+  version "10.4.2"
+  sha256 "2cac09e799f2eb609a7f1e4082d3d9d3d4d9e1930a8c4f9ecdad72aad35cdf10"
+
+  depends_on "cmake" => :build
+  depends_on "xerces-c"
+  depends_on "qt"
+
+  resource "G4NDL" do
+    url "http://cern.ch/geant4-data/datasets/G4NDL.4.5.tar.gz"
```
Comment (vszakats) [inline Formula/geant4.rb]: All `cern.ch` URLs support HTTPS, would mind updating them?

File: Formula/geant4.rb [hunk 6]
```diff
@@ -0,0 +1,100 @@
+class Geant4 < Formula
+  desc "Simulation toolkit for particle transport through matter"
+  homepage "http://geant4.cern.ch"
```
Comment (vszakats) [inline Formula/geant4.rb]: It'd be better to use this secure alternative:
https://geant4.web.cern.ch/

File: Formula/geant4.rb [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  url "http://geant4.cern.ch/support/source/geant4.10.04.p02.tar.gz"
```
Comment (vszakats) [inline Formula/geant4.rb]: Here's the secure alternative for the above URL:
https://geant4-data.web.cern.ch/geant4-data/releases/source/geant4.10.04.p02.tar.gz

File: Formula/geant4.rb (hunk 5 above)
Comment (henryiii) [reply Formula/geant4.rb]: Done. Forgot to check to see if they had secure versions, the old formula used http too.

File: Formula/geant4.rb (hunk 7 above)
Comment (henryiii) [reply Formula/geant4.rb]: Thanks!

**Comment by vszakats** [comment]
Looks good, thank you @henryiii!

**Comment by henryiii** [comment]
Dumb question: How do I add cmake as both a build and test dependency?

**Comment by fxcoudert** [comment]
@henryiii `depends_on "cmake" => [:build, :test]`

File: Formula/geant4.rb (hunk 3 above)
Comment (fxcoudert) [reply Formula/geant4.rb]: Move to install, please.

Comment (fxcoudert) [reply Formula/geant4.rb]: Also: are these datasets required to make the software useful? I'm not sure why a simulation toolkit would require large datasets, except as examples or tests

**Comment by fxcoudert** [comment]
```
  NoMethodError: undefined method `assert_in' for #<Formulary::FormulaNamespace16ea9529fc6b36b648aac8ba1985564f::Geant4:0x00000001014e6a68>
```

**Review by fxcoudert — APPROVED** [review]

**Pull request closed without merging**